### PR TITLE
Segment audio with Silero VAD and pipeline with Transcriber

### PR DESCRIPTION
**Pull request by am831**
Implements Silero VAD for segmenting long audio in the SileroVADSegmenter class and pipelines the class with the transcriber. When creating a transcriber object, the user can specify segment chunk size and pause length. By default, segment chunk size is 20 seconds and audio is automatically segmented if it is longer than the chunk size. 
Instead of using Silero's built in functions for segmenting (which are unable to guarantee segment length), this implementation uses a probabilistic divide and conquer algorithm to guarantee that segments do not exceed the max segment length. A threshold value is also used to filter out windows of audio that contain a low probability of speech.

Segmentation performance was evaluated on Silero VAD and Pydub with seamless. Silero performed slightly better so was selected for implementation. The evaluation was performed on a randomly sampled chunk of GigaSpeech under different conditions.

About the dataset:

```
Total hours of audio: 3.6
Mean audio duration: .3 hours
Audio files were selected randomly from the first 150 GB of GigaSpeech
```

Evaluation of Silero using probabilistic divide and conquer algorithm WITHOUT probability threshold filtering, 1 second pause length and 10 second max segment length:

```
Average CER: 0.69
Average Edit Distance: 12603.0

Mean segment duration: 5.9
Max segment duration: 7.9
Mean number of segments per audio sample: 182.6
```

Evaluation of Silero using probabilistic divide and conquer algorithm WITH probability threshold filtering, 1 second pause length and 10 second max segment length:

```
Average CER: 0.69
Average Edit Distance: 11894.5

Mean segment duration: 6.7
Max segment duration: 9.9
Mean number of segments per audio sample: 141.5
```

Evaluation of Silero using Silero's segmentation algorithm (previous commits), 1 second pause length and 10 second max segment length:

```
Average CER: 0.67
Average edit distance: 11370.3

Segmentation statistics:
Mean segment duration: 7.7 sec
Max segment duration: 107.6 sec
Mean number of segments per audio sample: 124.3
```

Evaluation of Pydub segmentation with 1 second pause length and 10 second max segment length: 

```
Average CER: 0.69
Average edit distance: 11642.9

Segmentation statistics:
Mean segment duration: 9.4 sec
Max segment duration: 10.0 sec
Mean number of segments per audio sample: 114.1

```

You can use the colab notebook I created for evaluation here: https://colab.research.google.com/drive/1KCRlGnUfKU5_T8bl_YFo_bf7ZawoMb6n?usp=sharing

Use this example for manual testing:
```
import torch

from seamless_communication.inference import Transcriber

model_name = "seamlessM4T_v2_large"
vocoder_name = "vocoder_v2" if model_name == "seamlessM4T_v2_large" else "vocoder_36langs"

transcriber = Transcriber (
    model_name,
    device=torch.device("cpu"),
    dtype=torch.float32,
)

input_audio = "example.wav"

txt = transcriber.transcribe(audio=input_audio, src_lang="eng")

print("Translated text: ", txt)
print()
```

### CODE REVIEW

File: src/seamless_communication/segment/silero_vad.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+# MIT_LICENSE file in the root directory of this source tree.
+
+from __future__ import annotations
+
+from argparse import Namespace
+import torch
+
+
+class SileroVADSegmenter:  # type: ignore
+    def __init__(self, args: Namespace) -> None:
```
Comment (avidale) [inline src/seamless_communication/segment/silero_vad.py:14]: I wonder why you provide a Namespace instead of more traditional keyword args?
Is there any advantage in doing so?

File: src/seamless_communication/segment/silero_vad.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        self.model, utils = torch.hub.load(
+            repo_or_dir="snakers4/silero-vad",
+            model="silero_vad",
+            force_reload=False,
+            onnx=False,
+        )
+
+        (
+            self.get_speech_timestamps,
```
Comment (avidale) [inline src/seamless_communication/segment/silero_vad.py:23]: I suggest that you re-implement the function `get_speech_timestamps` to something like PDAC (https://github.com/facebookresearch/stopes/blob/5f9cca5f36ec84476599eb5a47101b1dccef646a/stopes/modules/speech/shas/shas.py#L150) where we apply additional splitting to the segments that exceed the maximal length based on some softened criteria (either shorter pause duration, or larger threshold for predicted speech probability to consider it as non-pause).

File: src/seamless_communication/segment/silero_vad.py (hunk 1 above)
Comment (am831) [reply src/seamless_communication/segment/silero_vad.py:18]: Just to prevent users from passing unexpected arguments, but it isn't necessary here. It would also work to use kwargs or function parameters, my intention was to maintain style consistency with similar agent classes such as SileroVADAgent

File: src/seamless_communication/inference/transcriber.py [hunk 3]
```diff
@@ -316,10 +328,40 @@ def transcribe(
...
+              pause_length=pause_length_sec,
+              )
+            segmented_audios = segmenter.segment_long_input(waveform_1d)
+            transcriptions = []
+            for start, end in segmented_audios:
+                segment = waveform_2d[start:end, :]
+                src_segment = self.convert_to_fbank(
+                    {"waveform": segment, "sample_rate": decoded_audio.get("sample_rate"), 
+                     "format": decoded_audio.get("format")})["fbank"]
+                print(src_segment.device)
```
Comment (mavlyutovr) [inline src/seamless_communication/inference/transcriber.py:347]: remove

File: src/seamless_communication/segment/silero_vad.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+            min_segment_length: float, 
+            window_size_samples: float,
+        ) -> tp.List[Segment]:
+        """
+        Recursively splits segments based on speech threshold and duration. 
+        """
+        segments = []
+        sgm = Segment(0, len(probs)*window_size_samples, probs)
+
+        def recursive_split(sgm):
```
Comment (mavlyutovr) [inline src/seamless_communication/segment/silero_vad.py:108]: can you move it into a classmethod or a separate function and then add a unit test for it?

**Review by mavlyutovr — APPROVED** [review]

File: src/seamless_communication/inference/transcriber.py [hunk 5]
```diff
@@ -316,10 +328,40 @@ def transcribe(
...
+                    length_seconds_segment,
+                    filter_width,
+                    sequence_generator_options,
+                )
+                transcriptions.append(str(transcription_segment))
+
+            return " ".join(transcriptions)
+        else:
+
+            return self.run_inference(
```
Comment (mavlyutovr) [inline src/seamless_communication/inference/transcriber.py:361]: nit: to avoid having two copies of this call, can you implement this function differently: 
1) materialized src_sgments (for short audios it will be an array with one element - the original audio)
2) a loop with calls for each src_segment
3) fainal join

File: src/seamless_communication/inference/transcriber.py (hunk 3 above)
Comment (am831) [reply src/seamless_communication/inference/transcriber.py]: fixed

File: src/seamless_communication/segment/silero_vad.py [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+
+        if len(audio.shape) > 1:
+            for _ in range(audio.ndim):  # trying to squeeze empty dimensions
+                audio = audio.squeeze(0)
+            assert (
+                audio.ndim == 1
+            ), "More than one dimension in audio. Are you trying to process audio with 2 channels?"
+
+        if sampling_rate > 16000 and (sampling_rate % 16000 == 0):
+            step = sampling_rate // 16000
```
[the remark this answers is not in the document]
Comment (am831) [reply src/seamless_communication/segment/silero_vad.py]: fixed

File: src/seamless_communication/inference/transcriber.py (hunk 5 above)
Comment (am831) [reply src/seamless_communication/inference/transcriber.py]: fixed

**Review by mavlyutovr — APPROVED** [review]
